zerver/actions/message_flags.py

```diff
@@ -252,16 +253,22 @@ def do_update_message_flags(
     create_historical_user_messages(user_id=user_profile.id, message_ids=historical_message_ids)
 
     if operation == "add":
-        count = msgs.update(flags=F("flags").bitor(flagattr))
+        msgs = msgs.extra(where=[UserMessage.where_flag_is_absent(flag_int)])
+        updated_message_ids = [um.message_id for um in msgs]
+        count = len(updated_message_ids)
+        msgs.update(flags=F("flags").bitor(flagattr))
```

Comment:
I have to calculate `updated_message_ids` since after the `msgs.update` action, `msgs` length is `0`. I think this is due to the `extra where` query here, right?

Reply:
That's correct -- Django query objects are evaluated lazily. So it's doing a second query for the `.update` command from the one used to generated `updated_message_ids`.

The result is that this introduces a race, where those queries see a different set of unread messages. We could fix that with a transaction, but it would work just as well to replace the `msgs.update()` call with a `UserMessage.objects.filter(message_id__in=updated_message_ids, user_profile=user_profile).update(...)` query, so that we're changing the state of exactly the set of user IDs that we're going to report to clients were changed.

@mateuszmandera thoughts on this detail?

Reply:
#19054 is relevant here.  The logic in this function (and parallel logic in other functions in this file) is responsible for 100% of the deadlocks we see in production right now.

[Django doesn't support `UPDATE ... RETURNING id`](https://code.djangoproject.com/ticket/28682), frustratingly.  But doing this in two steps -- a transaction, then `SELECT id ... ORDER BY id FOR UPDATE`, then `UPDATE WHERE id in ...` would nicely address both of the issues, I think.

Reply:
What about doing this change, @amanagr?

Reply:
Updated, missed it somehow.

Reply:
> [Django doesn't support `UPDATE ... RETURNING id`](https://code.djangoproject.com/ticket/28682), frustratingly.

Ugh, and it's wontfix. :cry:

Reply:
I do note that that's purely because of return code reasons, and 5 years old, and maybe "add a flag which changes the return value" or something would get it accepted.  But yeah, frustrating.